OpenRA.Game/FileFormats/VocLoader.cs

```diff
@@ -0,0 +1,371 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+
+namespace OpenRA.FileFormats
+{
+	public class VocLoader : ISoundLoader
+	{
+		bool ISoundLoader.TryParseSound(Stream stream, string fileName, out byte[] rawData, out int channels, out int sampleBits, out int sampleRate)
+		{
+			try
+			{
+				var vocStream = new VocStream(stream);
+				channels = vocStream.Channels;
+				sampleBits = vocStream.BitsPerSample;
+				sampleRate = vocStream.SampleRate;
+				rawData = vocStream.ReadAllBytes();
+			}
+			catch
+			{
+				rawData = null;
+				channels = sampleBits = sampleRate = 0;
+				return false;
+			}
+
+			return true;
+		}
+	}
+
+	public class VocStream : Stream
+	{
+		public int BitsPerSample { get { return 8; } }
+		public int Channels { get { return 1; } }
+		public int SampleRate { get; private set; }
+
+		int totalSamples = 0;
+		int samplePosition = 0;
+
+		Stream stream;
+		List<VocBlock> blocks = new List<VocBlock>();
+		IEnumerator<VocBlock> currentBlock;
+		int samplesLeftInBlock = 0;
+		byte[] buffer = new byte[4096];
+
+		struct VocFileHeader
+		{
+			public string Description;
+			public int DatablockOffset;
+			public int Version;
+			public int ID;
+
+			public static VocFileHeader Read(Stream s)
+			{
+				VocFileHeader vfh;
+				vfh.Description = s.ReadASCII(20);
+				vfh.DatablockOffset = s.ReadUInt16();
+				vfh.Version = s.ReadUInt16();
+				vfh.ID = s.ReadUInt16();
+				return vfh;
+			}
+		}
+
+		struct VocBlock
+		{
+			public int Code;
+			public int Length;
+			public VocSampleBlock SampleBlock;
+			public VocLoopBlock LoopBlock;
+		}
+
+		struct VocSampleBlock
+		{
+			public int Rate;
+			public int Samples;
+			public long Offset;
+		}
+
+		struct VocLoopBlock
+		{
+			public int Count;
+		}
+
+		public VocStream(Stream stream)
+		{
+			this.stream = stream;
+			CheckVocHeader();
+			Preload();
+		}
+
+		private void CheckVocHeader()
+		{
+			var vfh = VocFileHeader.Read(stream);
+
+			if (!vfh.Description.StartsWith("Creative Voice File"))
+				throw new InvalidDataException("Voc header description not recognized");
+			if (vfh.DatablockOffset != 26)
+				throw new InvalidDataException("Voc header offset is wrong");
+			if (vfh.Version != 0x010A)
+				throw new InvalidDataException("Voc header version not recognized");
+			if (vfh.ID != ~vfh.Version + 0x1234)
+				throw new InvalidDataException("Voc header id is bogus");
```

Comment:
I understand that, but I'm asking you put them in the message so users can know what is wrong instead of having to come to us and then we redirect them to that page or another with documentation.